hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -239,10 +240,45 @@ def __stoppable_test_function(self, data):
                     # correct engine.
                     raise
 
-    def ir_tree_to_data(self, ir_tree_nodes):
+    def _cache(self, data):
+        result = data.as_result()
+        # when we shrink, we try out of bounds things, which can lead to the same
+        # data.buffer having multiple outcomes. eg data.buffer=b'' is Status.OVERRUN
+        # in normal circumstances, but a data with
+        # ir_nodes=[integer -5 {min_value: 0, max_value: 10}] will also have
+        # data.buffer=b'' but will be Status.INVALID instead.
+        #
+        # I think this indicates that we should be mark_overrun internally in
+        # these cases instead? alternatively, we'll map Status.INVALID to Overrun
+        # when caching.
+        if data.invalid_at is not None:
+            assert data.status is Status.INVALID
+            result = Overrun
+        self.__data_cache[data.buffer] = result
+        self.__data_cache_ir[tuple(data.examples.ir_tree_nodes)] = result
```

Comment:
If we cache ir nodes when we try a buffer, that may block future improvements for that set of ir nodes. eg we cache buffer A, then shrink to B < A, but cannot shrink to C < B < A because C corresponds to the same ir nodes as A.

This happens in practice here:

```
sets(integers(), min_size=3, max_size=3)
A: \x00\x00\x00\x00\x00\x02\x01\x00\x03 -> {0, 1, -1} ir: [0, 1, -1]
B: \x00\x00\x00\x00\x00\x02\x00\x00\x05 -> {0, 1, -2} ir: [0, 1, -2]
C: \x00\x00\x00\x00\x00\x02\x00\x00\x03 -> {0, 1, -1} ir: [0, 1, -1]
```

I didn't look into why the interpretation of \x01 vs \x00 here results in the same ir tree. The reason A gets tried as a buffer before C is dfa_replacement. Commenting out dfa_replacement shrinks this as normal.

I think this may be an acceptable intermediary tradeoff? It's unclear how much caching the corresponding ir for a buffer helps, frankly, but it's surely not zero.

Reply:
I'm considering killing two birds with one stone and removing some dfas here, which will have to happen / be converted to the ir eventually. I'm not seeing great test coverage over the benefits of these shrinking dfas in practice; is it enough if the test suite passes with good shrinking call numbers after removing dfas, or is additional manual testing in order? 

My guess is we can remove the floats and text dfas, but I'm not sure about emails or datetimes.

Reply:
I've sent @DRMacIver a message; it seems likely to me that if the tests pass without the DFAs then we probably don't need them in a post-IR world, or at least not in the current form.

Reply:
It does seem plausible to me that having IR would naturally solve some of the problems that DFAs were introduced to handle.

Reply:
I think it's fine to ditch the DFA logic. It was a clever idea that didn't get enough development to really prove its worth, and if it's holding back the IR switchover then I don't think it's super important to keep. It will probably cause some shrinking regressions but shrinking regressions are not the end of the world.

Reply:
Sounds good, thanks to both of you! Tests look clean after disabling `dfa_replacement` (d81d50e58fec06dbadfcac013c7fc8b47aeed29d), and the improvement in shrinking count is not insignificant either. `test_removes_needless_steps` is failing now but I think this is incidental since it doesn't explicitly fall under a dfa usecase.

I do think the DFA idea in general is cool and worth revisiting in the future.